### PR TITLE
Maybe save mesh pickle

### PR DESCRIPTION
**Pull request by pgierz**
Hi Nikolay,

Gregor ran into a problem using pfinterp because he didn't have permission to save the pickle file. I added a `try/except` around that, so it now only will save the file if you have permission -- otherwise, the entire process crashes. Now at least, you always get the mesh object back.

### CODE REVIEW

**Comment by pgierz** [comment]
Not sure what the failed check is. I looked and it seems to install a miniconda env to do it's test in, and from the output that looks fine but then it just stops...

**Comment by koldunovn** [comment]
Thanks, Paul! It make sense. I will try to fix the testing now (looks like something with installation), and will merge when everything is back online :)

**Comment by pgierz** [comment]
Nikolay, maybe don't put the button just yet. Gregor just sent me another error, I think I need some more try's somewhere:
```
PermissionError: [Errno 13] Permission denied: '/work/ollie/dsidoren/input/fesom2.0/meshes/mesh_CORE2_finaltopo_mean/distances_126858_-180.0_180.0_-80.0_90.0_360_170_1'
```

**Comment by pgierz** [comment]
OK, I found where the other change needs to happen.

Design question -- I can write this so that it tries to write to the mesh path, if that fails, it tries to write to some other path (maybe an env variable `PYFESOM_CACHE`, which defaults to `${HOME}/.cache/pyfesom`), and if that also fails, it doesn't write anything but still gives you back the regrid weights/mesh object.

**Comment by koldunovn** [comment]
> OK, I found where the other change needs to happen.
> 
> Design question -- I can write this so that it tries to write to the mesh path, if that fails, it tries to write to some other path (maybe an env variable `PYFESOM_CACHE`, which defaults to `${HOME}/.cache/pyfesom`), and if that also fails, it doesn't write anything but still gives you back the regrid weights/mesh object. Loading would happen in the same order -- try to read binary files from mesh, then from cache, and then just brutally recalculate the grid weights.

I would go for mesh path first, then `./` and then warning and then return interpolated.
In the `fesom2regular` there is a `basepath` variable, that should make it easier to implement. Thanks for doing this!

**Comment by pgierz** [comment]
OK, just finished. Have a look and test it out; I don't have any test data at hand right now.

File: pyfesom2/regridding.py [hunk 1]
```diff
@@ -128,40 +128,87 @@ def fesom2regular(
...
-            inds = joblib.load(inds_path)
-        else:
+        for distances_path in distances_paths:
+            if os.path.isfile(distances_path):
+                logging.info(
+                    "Note: using precalculated file from {}".format(distances_path)
+                )
+                try:
+                    distances = joblib.load(distances_path)
+                    loaded_distances = True
```
Comment (pgierz) [inline pyfesom2/regridding.py:180]: missing a break here

File: pyfesom2/regridding.py [hunk 2]
```diff
@@ -128,40 +128,87 @@ def fesom2regular(
...
+                except PermissionError:
+                    # who knows, something didn't work. Try the next path:
+                    continue
+        for inds_path in inds_paths:
+            if os.path.isfile(inds_path):
+                logging.info("Note: using precalculated file from {}".format(inds_path))
+                try:
+                    inds = joblib.load(inds_path)
+                    loaded_inds = True
```
Comment (pgierz) [inline pyfesom2/regridding.py:189]: Here too

File: pyfesom2/regridding.py [hunk 3]
```diff
@@ -170,20 +217,45 @@ def fesom2regular(
...
-            inds = joblib.load(inds_path)
-        else:
+        for distances_path in distances_paths:
+            if os.path.isfile(distances_path):
+                logging.info(
+                    "Note: using precalculated file from {}".format(distances_path)
+                )
+                try:
+                    distances = joblib.load(distances_path)
+                    loaded_distances = True
```
Comment (pgierz) [inline pyfesom2/regridding.py:227]: Missing `break`

File: pyfesom2/regridding.py [hunk 4]
```diff
@@ -170,20 +217,45 @@ def fesom2regular(
...
+                except PermissionError:
+                    # who knows, something didn't work. Try the next path:
+                    continue
+        for inds_path in inds_paths:
+            if os.path.isfile(inds_path):
+                logging.info("Note: using precalculated file from {}".format(inds_path))
+                try:
+                    inds = joblib.load(inds_path)
+                    loaded_inds = True
```
Comment (pgierz) [inline pyfesom2/regridding.py:236]: Missing `break`

File: pyfesom2/regridding.py [hunk 5]
```diff
@@ -194,27 +266,53 @@ def fesom2regular(
...
-        else:
+        for qhull_path in qhull_paths:
+            if os.path.isfile(distances_path):
+                logging.info(
+                    "Note: using precalculated file from {}".format(distances_path)
+                )
+                logging.info("Note: using precalculated file from {}".format(qhull_path))
+                try:
+                    qh = joblib.load(qhull_path)
+                    loaded_qhull = True
```
Comment (pgierz) [inline pyfesom2/regridding.py:277]: Missing `break`

File: pyfesom2/regridding.py [hunk 6]
```diff
@@ -194,27 +266,53 @@ def fesom2regular(
...
```
Comment (pgierz) [inline pyfesom2/regridding.py:303]: Missing `break`

**Comment by koldunovn** [comment]
Paul, thanks a lot. I will try to fix the testing first and then will accept it.

**Comment by pgierz** [comment]
No worries, thanks for building the nice tools! Collaborative projects like this only work if everyone chips in ;-)

**Comment by koldunovn** [comment]
I replace the placement of the default cache to the current working directory for two reasons:
 - looking for HOME will break Windows compatibility
 - Putting it to .cache will lead to growth in the size of the HOME directory which is hard to track (at leas this is my experience).

There is still a chance that you do analysis in the place where you have no write permissions to, but I think we have to live with that. 

What do you think, @pgierz ?

**Comment by pgierz** [comment]
We have windows users?? 🤦 

Everything else is fine by me :-)

**Comment by koldunovn** [comment]
I don't know if we are, but tests are passing on Windows, and it would be nice to keep it this way :)

**Pull request merged**